File: P1/Ejercicio3/src/constructor/ConstructorBicicletaMontana.java
package constructor;

import bicicleta.BicicletaMontana;

/**
 * Clase constructora para bicicletas de montaña. Esta hereda de Constructor
 */
public class ConstructorBicicletaMontana extends Constructor {
    /**
     * Constructor de la clase sin parámetros. Inicializa la bicicleta a una bicicleta de montaña por defecto
     */
    public ConstructorBicicletaMontana() { bicicleta = new BicicletaMontana(); }

    /**
     * Construye la suspensión
     * @param tipo Tipo de suspensión
     * @param cantidad Cantidad de suspensiones
     */
    public void construirSuspension(String tipo, int cantidad) {
        ((BicicletaMontana) bicicleta).establecerSuspension(tipo, cantidad);
    }

    /**
     * Implementación del método abstracto 'recomponer', que reinicializa la instancia de bicicleta de montaña
     */
    @Override
    public void recomponer() {
        bicicleta = new BicicletaMontana();
    }
}

File: P2/lib/Constructores/constructorBicicletaMontana.dart
import 'constructor.dart';
import '../Bicicleta/BicicletaMontana.dart';

class ConstructorBicicletaMontana extends Constructor {
  ConstructorBicicletaMontana() {
    bicicleta = BicicletaMontana();
  }

  void construirSuspension(String tipo, int cantidad) {
    // Dart no permite la conversión de tipos como en Java. En su lugar,
    // puedes usar 'as' para castear, asegurándote primero de que no sea null.
    // Usa is para el instanceof
    if (bicicleta is BicicletaMontana) {
      (bicicleta as BicicletaMontana).establecerSuspension(tipo, cantidad);
    }
  }

  @override
  void recomponer() {
    bicicleta = BicicletaMontana();
  }
}
